components/gitpod-db/src/typeorm/auth-code-repository-db.ts

```diff
@@ -0,0 +1,84 @@
+ * See License-AGPL.txt in the project root for license information.
+ */
+
+import { log } from '@gitpod/gitpod-protocol/lib/util/logging';
+import { DateInterval, OAuthAuthCode, OAuthAuthCodeRepository, OAuthClient, OAuthScope, OAuthUser } from "@jmondi/oauth2-server";
+import * as crypto from 'crypto';
+import { inject, injectable } from "inversify";
+import { EntityManager, Repository } from "typeorm";
+import { DBOAuthAuthCodeEntry } from './entity/db-oauth-auth-code';
+import { TypeORM } from './typeorm';
+
+const expiryInFuture = new DateInterval("5m");
+
+@injectable()
+export class AuthCodeRepositoryDB implements OAuthAuthCodeRepository {
+
+    @inject(TypeORM)
+    private readonly typeORM: TypeORM;
+
+    protected async getEntityManager(): Promise<EntityManager> {
+        return (await this.typeORM.getConnection()).manager;
+    }
+
+    async getOauthAuthCodeRepo(): Promise<Repository<DBOAuthAuthCodeEntry>> {
+        return (await this.getEntityManager()).getRepository<DBOAuthAuthCodeEntry>(DBOAuthAuthCodeEntry);
+    }
+
+    public async getByIdentifier(authCodeCode: string): Promise<OAuthAuthCode> {
+        log.info(`getByIdentifier ${authCodeCode}`);
+        const authCodeRepo = await this.getOauthAuthCodeRepo();
+        let authCodes = await authCodeRepo.find({ code: authCodeCode });
+        log.info(`getByIdentifier pre: ${JSON.stringify(authCodes)}`);
+        authCodes = authCodes.filter(te => (new Date(te.expiresAt)).getTime() > Date.now());
+        log.info(`getByIdentifier post: ${JSON.stringify(authCodes)}`);
+        const authCode = authCodes.length > 0 ? authCodes[0] : undefined;
+        return new Promise<OAuthAuthCode>((resolve, reject) => {
+            if (authCode) {
+                log.info(`getByIdentifier found ${authCodeCode} ${JSON.stringify(authCode)}`);
+                resolve(authCode);
+            } else {
+                log.info(`getByIdentifier failed to find ${authCodeCode}`);
+                reject(`authentication code not found`);
+            }
+        });
+    }
+    public issueAuthCode(client: OAuthClient, user: OAuthUser | undefined, scopes: OAuthScope[]): OAuthAuthCode {
+        const code = crypto.randomBytes(30).toString('hex');
+        log.info(`issueAuthCode: ${JSON.stringify(client)}, ${JSON.stringify(user)}, ${JSON.stringify(scopes)}, ${code}`);
+        return {
+            code: code,
+            user,
+            client,
+            redirectUri: "",
+            codeChallenge: undefined,
+            codeChallengeMethod: undefined,
+            expiresAt: expiryInFuture.getEndDate(),
+            scopes: [],
+        };
+    }
+    public async persist(authCode: OAuthAuthCode): Promise<void> {
+        log.info(`persist auth ${JSON.stringify(authCode)}`);
+        const authCodeRepo = await this.getOauthAuthCodeRepo();
+        authCodeRepo.save(authCode);
+    }
+    public async isRevoked(authCodeCode: string): Promise<boolean> {
```

Comment:
Def [called](https://github.com/jasonraimondi/typescript-oauth2-server/blob/153c83bd64940a03ba31c833d4e3b2aeb4ce03cd/src/grants/auth_code.grant.ts#L231) by lib and is an important part of the required interface.

Agreed that it looks like isExpired because that is exactly what it is doing :-)
Given we are very unlikely to want to support long-lived auth tokens I didn't want to go to the trouble of extending the interface to add something more suitable. The intention is to add a cleanup phase to remove expired auth codes at a later date.